driver/src/orderbook/streamed/block_timestamp_reading.rs

```diff
@@ -22,34 +36,83 @@ impl BlockTimestampReading for Web3 {
     }
 }
 
-/// A cache for the block timestamp which avoids having to query the node in the case where we
-/// receive multiple events from the same block in a row.
+impl BlockTimestampBatchReading for Web3 {
+    fn block_timestamps(
+        &mut self,
+        block_hashes: HashSet<H256>,
+    ) -> BoxFuture<Result<Vec<(H256, u64)>>> {
+        let batched_web3 = ethcontract::web3::Web3::new(Batch::new(self.transport().clone()));
+        async move {
+            let mut result = Vec::with_capacity(block_hashes.len());
+            for chunk in Vec::from_iter(block_hashes.into_iter()).chunks(1000) {
+                let partial_result = query_block_timestamps_batched(&batched_web3, chunk).await?;
+                result.extend(partial_result);
+            }
+            Ok(result)
+        }
+        .boxed()
+    }
+}
+
+type BatchedWeb3 = ethcontract::web3::Web3<Batch<HttpTransport>>;
+async fn query_block_timestamps_batched(
+    batched_web3: &BatchedWeb3,
+    block_hashes: &[H256],
+) -> Result<Vec<(H256, u64)>> {
+    block_hashes.iter().for_each(|hash| {
+        batched_web3.eth().block(BlockId::Hash(*hash));
+    });
+    let result = batched_web3.transport().submit_batch().compat().await;
+    result
+        .with_context(|| "Batch RPC call to get block hashes failed")?
+        .into_iter()
+        .map(|response| {
+            let block: Block<H256> = serde_json::from_value(response?)?;
+            Ok((
+                block
+                    .hash
+                    .expect("blocks queried by hash should contain a hash"),
+                block.timestamp.low_u64(),
+            ))
+        })
+        .collect()
+}
+
+/// A cache for the block timestamp.
 #[derive(Debug)]
 pub struct MemoizingBlockTimestampReader<T> {
```

Comment:
naming nit: This should probably be called `CachedBlockTimestampReader` (I reviewed the original PR before the name change) as I would describe what this is doing as caching more than memoizing.